cmd/metal-api/internal/service/network-service.go

```diff
@@ -523,40 +630,72 @@ func (r *networkResource) allocateNetwork(request *restful.Request, response *re
 		DestinationPrefixes: destPrefixes,
 		Shared:              shared,
 		Nat:                 nat,
+		AddressFamilies:     superNetwork.AddressFamilies,
+	}
+
+	// Allow configurable prefix length per AF
+	length := superNetwork.DefaultChildPrefixLength
+	if len(requestPayload.Length) > 0 {
+		for af, l := range requestPayload.Length {
+			length[metal.ToAddressFamily(string(af))] = l
+		}
+	}
+
+	if requestPayload.AddressFamily != nil {
+		af := metal.ToAddressFamily(string(*requestPayload.AddressFamily))
+		bits, ok := length[af]
+		if !ok {
+			r.sendError(request, response, httperrors.BadRequest(fmt.Errorf("addressfamiliy %s specified, but no childprefixlength for this addressfamily", *requestPayload.AddressFamily)))
+			return
+		}
+		length = metal.ChildPrefixLength{
+			af: bits,
+		}
+		nwSpec.AddressFamilies = metal.AddressFamilies{af}
 	}
+
+	r.log.Info("network allocate", "supernetwork", superNetwork.ID, "defaultchildprefixlength", superNetwork.DefaultChildPrefixLength, "length", length)
+
 	ctx := request.Request.Context()
-	nw, err := createChildNetwork(ctx, r.ds, r.ipamer, nwSpec, &superNetwork, partition.PrivateNetworkPrefixLength)
+	nw, err := r.createChildNetwork(ctx, nwSpec, &superNetwork, length)
 	if err != nil {
 		r.sendError(request, response, defaultError(err))
 		return
 	}
 
-	usage := getNetworkUsage(ctx, nw, r.ipamer)
+	usage, err := r.getNetworkUsage(ctx, nw)
+	if err != nil {
+		r.sendError(request, response, defaultError(err))
+		return
+	}
 
 	r.send(request, response, http.StatusCreated, v1.NewNetworkResponse(nw, usage))
 }
 
-func createChildNetwork(ctx context.Context, ds *datastore.RethinkStore, ipamer ipam.IPAMer, nwSpec *metal.Network, parent *metal.Network, childLength uint8) (*metal.Network, error) {
-	vrf, err := acquireRandomVRF(ds)
+func (r *networkResource) createChildNetwork(ctx context.Context, nwSpec *metal.Network, parent *metal.Network, childLengths metal.ChildPrefixLength) (*metal.Network, error) {
+	vrf, err := acquireRandomVRF(r.ds)
 	if err != nil {
 		return nil, fmt.Errorf("could not acquire a vrf: %w", err)
 	}
 
-	childPrefix, err := createChildPrefix(ctx, parent.Prefixes, childLength, ipamer)
-	if err != nil {
-		return nil, err
-	}
-
-	if childPrefix == nil {
-		return nil, fmt.Errorf("could not allocate child prefix in parent network: %s", parent.ID)
+	var childPrefixes = metal.Prefixes{}
+	for af, childLength := range childLengths {
+		childPrefix, err := r.createChildPrefix(ctx, parent.Prefixes, af, childLength)
+		if err != nil {
+			return nil, err
+		}
+		if childPrefix == nil {
+			return nil, fmt.Errorf("could not allocate child prefix in parent network: %s for addressfamily: %s length:%d", parent.ID, af, childLength)
```

Comment:
Hard to mitigate this but this is a potential leakage in case one child prefix could already be allocated.

Reply:
Not quite sure what you mean. Cleaning up the created child prefixes that went smoothly?